runtime/coverage.go

```diff
@@ -20,56 +20,179 @@ package runtime
+	coveredLines := c.CoveredLines()
+	percentage := 100 * float64(coveredLines) / float64(c.Statements)
+	c.Percentage = fmt.Sprintf("%0.1f%%", percentage)
 }
 
-func NewLocationCoverage() *LocationCoverage {
+func (c *LocationCoverage) CoveredLines() int {
+	coveredLines := 0
+	for _, hits := range c.LineHits { // nolint:maprange
+		if hits > 0 {
+			coveredLines += 1
+		}
+	}
+	return coveredLines
+}
+
+func (c *LocationCoverage) MissedLines() []int {
+	missedLines := make([]int, 0)
+	for line, hits := range c.LineHits { // nolint:maprange
+		if hits == 0 {
+			missedLines = append(missedLines, line)
+		}
+	}
+	return missedLines
+}
+
+func NewLocationCoverage(statementLines []int) *LocationCoverage {
+	lineHits := make(map[int]int, len(statementLines))
+	for _, line := range statementLines {
+		lineHits[line] = 0
+	}
 	return &LocationCoverage{
-		LineHits: map[int]int{},
+		LineHits:   lineHits,
+		Statements: len(lineHits),
+		Percentage: "0.0%",
 	}
 }
 
-// CoverageReport is a collection of coverage per location
+// CoverageReport is a collection of coverage information per location
 type CoverageReport struct {
-	Coverage map[common.Location]*LocationCoverage `json:"-"`
+	Coverage          map[common.Location]*LocationCoverage `json:"-"`
+	Programs          map[common.Location]*ast.Program      `json:"-"`
+	ExcludedLocations []common.Location                     `json:"-"`
+}
+
+func (r *CoverageReport) ExcludeLocation(location Location) {
+	if r.IsLocationExcluded(location) {
+		return
+	}
+	r.ExcludedLocations = append(r.ExcludedLocations, location)
+}
+
+func (r *CoverageReport) IsLocationExcluded(location Location) bool {
+	for _, excludedLocation := range r.ExcludedLocations {
+		if excludedLocation == location {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *CoverageReport) AddLineHit(location Location, line int) error {
+	if r.IsLocationExcluded(location) {
+		return fmt.Errorf("location is excluded from coverage: %v", location)
+	}
+
+	if !r.IsProgramInspected(location) {
+		return fmt.Errorf("program is not inspected for location: %v", location)
+	}
+
+	locationCoverage := r.Coverage[location]
+	locationCoverage.AddLineHit(line)
+
+	return nil
 }
 
-func (r *CoverageReport) AddLineHit(location Location, line int) {
+func (r *CoverageReport) InspectProgram(location Location, program *ast.Program) error {
+	if r.IsLocationExcluded(location) {
+		return fmt.Errorf("location is excluded from coverage: %v", location)
+	}
+	r.Programs[location] = program
+	var statements []int
+
+	inspector := ast.NewInspector(program)
+	inspector.Elements(
+		[]ast.Element{
+			(*ast.ReturnStatement)(nil),
+			(*ast.BreakStatement)(nil),
+			(*ast.ContinueStatement)(nil),
+			(*ast.IfStatement)(nil),
+			(*ast.WhileStatement)(nil),
+			(*ast.ForStatement)(nil),
+			(*ast.EmitStatement)(nil),
+			(*ast.AssignmentStatement)(nil),
+			(*ast.SwapStatement)(nil),
+			(*ast.ExpressionStatement)(nil),
+			(*ast.VariableDeclaration)(nil),
```

Comment:
There is indeed a reason for filtering specific AST element types, but correct if I am wrong.

The [newOnStatementHandler](https://github.com/onflow/cadence/blob/master/runtime/environment.go#L613-L618) is not called with all `ast.Statement` elements found on a program/location.

Hence, we will often have missed lines for statements that are not meant to be executed.

I have created a [gist](https://gist.github.com/m-Peter/980bbe50f235b89e89a027f8a2ec9475) which attempts to show what  I explained above.

The `"missed_lines": [1, 4, 13, 17]` from the `coverage.json` file, correspond to the following `ast.Statement`:
- ElementType: `ElementTypeCompositeDeclaration`, Statement: `*ast.CompositeDeclaration`
- ElementType: `ElementTypeSpecialFunctionDeclaration`, Statement: `*ast.SpecialFunctionDeclaration`
- ElementType: `ElementTypeFunctionDeclaration`, Statement: `*ast.FunctionDeclaration`
- ElementType: `ElementTypeFunctionDeclaration`, Statement: `*ast.FunctionDeclaration`

Is there any other way around this?

Reply:
Thank you for the detailed explanation, it totally makes sense now what the "problem" is: some `ast.Statement` are `ast.Declaration`, e.g `ast.FunctionDeclaration`, because they may appear as a "statement" inside of a function body – but we do not expect all of them to be executed/interpreted (and get a statement handler callback), but only some (e.g. `ast.VariableDeclaration`).

Manually specifying this particular set here is brittle, as additional statements might get added in the future, which means this set has to be maintained.

It might be better to look for any `ast.Statement`, so that all statements are handled, but also filter ones that are `ast.Declaration` – except for one  (for now) particular declaration, which is executed/interpreted, `VariableDeclaration`.

Reply:
Now it's quite clear :pray: I also observed this with some trial & error. For example:

```cadence
pub let specialNumbers: {Int: String} = {
    1729: "Harshad",
    8128: "Harmonic",
    41041: "Carmichael"
}
```
would not be executed/interpreted, even though it is an `ast.VariableDeclaration`. That's because it is declared on the top-level.

However,

```cadence
for input in testInputs.keys {
    let result = getIntegerTrait(input)
    ...
}
```

this would be executed/interpreted, because it is declared inside the `main()` function.

I have updated the the logic of the `InspectProgram` method here: https://github.com/onflow/cadence/pull/2296/commits/4cc2c02e21bb7491ca708d4b75c03579a8e86c87.
Let me know if I got the logic straight, and if there's a more elegant way to track which `ast.VariableDeclaration` are locally declared.

Reply:
Nice refactor, the `isStatement && !isDeclaration` logic is a bit more future proof 👍 

Regarding detecting non-top level variable declarations: Using the start position of the variable declaration does not handle a program which does not use indentation (bad style, but still valid). 

You can improve this by using the `push` parameter to keep a depth variable. At depth == 2 (program + variable declaration), you know you are inspecting a top-level declaration vs a local variable declaration:

```go
func TestInspectorNonTopLevelVariableDeclarations(t *testing.T) {

	t.Parallel()

	const code = `
      let x = 1

      fun f() {
	      let y = 2
      }

      let z = 3
    `

	program, err := parser.ParseProgram(
		nil,
		[]byte(code),
		parser.Config{},
	)
	require.NoError(t, err)

	var depth int
	var topLevelVariables int
	var localVariables int

	inspector := ast.NewInspector(program)
	inspector.Elements(
		nil,
		func(element ast.Element, push bool) bool {
			if push {
				depth++

				if _, ok := element.(*ast.VariableDeclaration); ok {
					// 2: program + variable declaration
					atTopLevel := depth == 2

					if atTopLevel {
						topLevelVariables += 1
					} else {
						localVariables += 1
					}
				}
			} else {
				depth--
			}

			return true
		},
	)

	assert.Equal(t, 2, topLevelVariables)
	assert.Equal(t, 1, localVariables)
}
```

Reply:
Arigato, sensei :bow:
That was a perfect explanation on the defects of using the start position :100:
Not sure if it would cross my mind :sweat_smile: